src/layer/unitary.rs

```diff
@@ -0,0 +1,465 @@
+use num;
+use af;
+use af::{Array, MatProp, Dim4, DType};
+
+use std::sync::{Arc, Mutex};
+use activations;
+use initializations;
+use params::{Params};
+use layer::Layer;
+
+use num::Complex;
+
+
+pub struct Unitary {
+    pub input_size: usize,
+    pub output_size: usize,
+}
+
+fn h_d(param: Array, ar: Array) -> Array {
+    let D = af::cplx2(&af::cos(&param)
+                      , &af::sin(&param)
+                      ,false);
+    af::mul(&D, &ar, true)
+}
+
+fn r_d(param: Array, ar: Array) -> Array {
+    let D = af::cplx2(&af::cos(&param)
+                      , &af::mul(&af::sin(&param), &-1, true)
+                      ,false);
+    af::mul(&D, &ar, true)
+}
+
+fn h_fft(ar: Array) -> Array {
+        let ar_size = ar.dims()[1];
+        af::transpose(&af::fft(&af::transpose(&ar, false), 1.0, ar_size as i64)
+                      , false)
+}
+
+fn r_fft(ar: Array) -> Array {
+        let ar_size = ar.dims()[1];
+        af::transpose(&af::ifft(&af::transpose(&ar, false), 1.0, ar_size as i64)
+                      , false)
+}
+
+fn h_ifft(ar: Array) -> Array {
+        let ar_size = ar.dims()[1];
+        af::transpose(&af::ifft(&af::transpose(&ar, false), 1.0/(ar_size as f64), ar_size as i64)
+                              , false)
+}
+
+fn r_ifft(ar: Array) -> Array {
+        let ar_size = ar.dims()[1];
+        af::transpose(&af::fft(&af::transpose(&ar, false), 1.0/(ar_size as f64), ar_size as i64)
+                              , false)
+}
+
+fn h_pi(param: Array, ar: Array) -> Array {
+        af::lookup(&ar, &param, 1)
+}
+
+fn h_r(param: Array, mut ar: Array) -> Array {
+        let ar_temp = ar.clone();
+        ar = af::matmul(&param
+                       , &ar
+                       , MatProp::NONE
+                       , MatProp::TRANS);
+        ar = af::matmul(&ar
+                       , &af::conjg(&param)
+                       , MatProp::TRANS
+                       , MatProp::NONE);
+        ar = af::mul(&ar, &2, true);
+        ar = af::sub(&ar_temp, &ar, false);
+        ar
+
+}
+
+// Wh = (D3 R2 F-1 D2 Pi R1 F D1) h
+fn wh(p1: Array, p2: Array, p3: Array, p4: Array, p5: Array, p6: Array, ar: Array) -> Array { 
+    let mut current = ar;
+    current = h_d(p1, current);
+    current = h_fft(current);
+    current = h_r(p2, current);
+    current = h_pi(p3, current);
+    current = h_d(p4, current);
+    current = h_ifft(current);
+    current = h_r(p5, current);
+    current = h_d(p6, current);
+    current
+           
+}
+
+fn to_complex(ar:Array) -> Array {  
+    let dim = ar.dims()[1];
+    af::cplx2(&af::cols(&ar, 0, dim/2-1), &af::cols(&ar, dim/2, dim-1), false)
+}
+
+fn to_real(ar: Array) -> Array {
+    af::join(1, &af::real(&ar), &af::imag(&ar))
+}
+
+
+impl Layer for Unitary
+{
+    fn forward(&self, params:  Arc<Mutex<Params>>, inputs: &Array, state: Option<Array>) -> (Array, Option<Array>)
```

Comment:
Btw: if you look at the latest RNN code, if `state` is provided it utilizes that state (no matter if it's 0th state or not). This is also helpful during the gradient test. I.e. we provide some random state here.